src/data_availability.rs

```diff
@@ -233,6 +341,89 @@ impl DataAvailability {
                 hash,
             }));
     }
+
+    /// API endpoint to request a stream of blocks from a peer.
+    /// The DA module will start at the requested block
+    /// or its oldest stored block if the requested block is not present.
+    /// New blocks will be pushed as soon as they are received.
+    /// Blocks may arrive out of order and duplicates are possible.
+    async fn api_stream_request(
+        Query(filters): Query<StreamRequest>,
+        State(state): State<tokio::sync::mpsc::Sender<StreamRequest>>,
+    ) -> Result<(), StatusCode> {
+        if let Err(err) = state.send(filters).await {
+            error!("Error passing stream request to DA: {}", err);
+            return Err(StatusCode::INTERNAL_SERVER_ERROR);
+        }
+        Ok(())
+    }
+
+    /// Process a stream request from a peer.
+    async fn handle_stream_request(&mut self, start_height: u64, peer_ip: &str) -> Result<()> {
+        let stream = TcpStream::connect(peer_ip).await?;
+        info!("DA stream setup on {} to {}", stream.local_addr()?, peer_ip);
+        let (sender, mut receiver) = Framed::new(stream, LengthDelimitedCodec::new()).split();
+
+        // Start a task to process pings from the peer.
+        // We do the processing in the main select! loop to keep things synchronous.
+        let peer_ip_keepalive = peer_ip.to_string();
+        let ping_sender = self.ping_sender.clone();
+        let keepalive_abort = tokio::spawn(async move {
+            loop {
+                receiver.next().await;
+                let _ = ping_sender.send(peer_ip_keepalive.clone()).await;
```

Comment:
Not entirely which logic you talk about, I guess the `peer_ip_keepalive` handling?
If so, I did it that way because I borrow `self` mutably and this wouldn't really play well with rust structures otherwise.

The design is a little future-looking in that currently we use TCP so I could juste remove this code altogether as an alternative.

Reply:
~my point was more on the fact that you send a message on the ping_sender channel when you get data from the stream~

Ok I just understood, here on the stream, the only frames you get are pings... I thought that it could be other stuff too, I'm starting to understant this file :)